### PR TITLE
Host firewall fixes

### PR DESCRIPTION
**Pull request by pchaigno**
This pull request fixes various host firewall bugs. Please see the release notes and review commit by commit.

### How I Tested This

I'm still tracking some bugs in the host firewall, so seems important to highlight what I tested and is known to work.

I tested the host firewall with the [ingress+egress L3+L4 example policy](https://github.com/cilium/cilium/blob/pr/pchaigno/host-firewall-fixes/examples/policies/host/lock-down-dev-vms.yaml) included at the end of this series. It attempts to whitelist connections in the most fine-grained way possible. Specifically, I launched the connectivity-test and loaded that policy in the net-next development VMs, with options `kube-proxy-replacement=strict`, `enable-remote-node-identity`, `enable-host-firewall`, and the devices set.

I then checked that the connectivity check was okay for 20min+ and verified all packet drops were expected. 

### Impact on CI

The host firewall is enabled in our Packet K8s CI builds. The main reason to enable it was to be able to detect any regression due to BPF complexity issues (it adds a lot of code to the datapath). That was fine because having the host firewall enabled with allow-all default shouldn't impact anything in Cilium (apart from dropping packets if there's a bug).

However, with the present fixes, enabling the host firewall now changes the path of some packets to force them through the firewall's ingress. That means we are really not testing the default setup (i.e., host firewall disabled) the users are likely to have.

I will therefore send a follow up PR to only enable the host firewall in CI if we have some label set in the PR (as we do for other features). 

Updates: #11799

```release-note
Fix incorrect host firewall enforcement when used with BPF-based NodePort services
Fix host firewall ingress bypass on path from pods to local host
Fix potential ingress host firewall bypass in tunneling mode for remote pods
Fix handling of ICMPv6 messages by host firewall
Fix failure to recognize established IPv6 connections on egress of the host firewall
Fix potential host firewall drops on egress of the node in case of SNAT
```

### CODE REVIEW

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
retest-runtime

File: bpf/bpf_host.c
Comment (qmonnet) [inline bpf/bpf_host.c]: Nit: You likely lost this fix in a rebase:
```suggestion
		 * rewrite the destination MAC address to the MAC of cilium_net.
```

File: bpf/lib/icmp6.h
Comment (qmonnet) [inline bpf/lib/icmp6.h]: Also lost from a rebase:
```suggestion
/* If no specific action is specified, drop unknown neighbour solicitation
 * messages.
```

File: bpf/lib/icmp6.h
Comment (qmonnet) [inline bpf/lib/icmp6.h]: Maybe use a lower value? Although we are unlikely to ever use it, `TC_ACT_UNSPEC` is `-1`.

File: test/k8sT/Services.go
Comment (qmonnet) [inline test/k8sT/Services.go:1176]: Are those incompatibilities (this one, and with portmap chaining) documented somewhere?

**Review by qmonnet — COMMENTED** [review]
Aside from a few minor nits, the PR looks good to me. Thanks for this work!

**Review by nebril — APPROVED** [review]
CI bits lgtm

**Review by aanm — APPROVED** [review]

File: test/k8sT/Services.go
Comment (pchaigno) [reply test/k8sT/Services.go:1176]: Not yet. I have several reminders to create specific issues for those incompatibilities once the present PR is merged. I would prefer to avoid creating those issues before in case another approach is found through the reviews.

**Comment by pchaigno** [comment]
test-me-please

**Review by qmonnet — APPROVED** [review]

**Comment by brb** [comment]
Re `We want the host firewall to see ingress-to-host packets (1) after the
DNAT but before the SNAT on the way to the remote backend` from https://github.com/cilium/cilium/pull/12345/commits/69c4efe39dee5ab3a8101b4862fddcec9a0b8008: please correct me if I'm wrong, but I think we might still miss this case, as in the case of a remote backend, `CILIUM_CALL_IPV{4,6}_NODEPORT_NAT` (taill-called by `nodeport_lb{4,6}()`) will not recircle back to bpf_host and will invoke `ctx_redirect()`.

**Comment by pchaigno** [comment]
> Re `We want the host firewall to see ingress-to-host packets (1) after the DNAT but before the SNAT on the way to the remote backend` from [69c4efe](https://github.com/cilium/cilium/commit/69c4efe39dee5ab3a8101b4862fddcec9a0b8008): please correct me if I'm wrong, but I think we might still miss this case, as in the case of a remote backend, `CILIUM_CALL_IPV{4,6}_NODEPORT_NAT` (taill-called by `nodeport_lb{4,6}()`) will not recircle back to bpf_host and will invoke `ctx_redirect()`.

Ah, right. So if the backend is remote, I can just skip the ingress host policy enforcement. Is it correct to assume that if `CILIUM_CALL_IPV{4,6}_NODEPORT_NAT` is called with `NAT_DIR_EGRESS`, then the backend is remote? If that's always true, I could maybe set a mark to skip the host firewall (I'm guessing `ctx->tc_index` isn't preserved on redirects?).

**Comment by brb** [comment]
> Is it correct to assume that if CILIUM_CALL_IPV{4,6}_NODEPORT_NAT is called with NAT_DIR_EGRESS, then the backend is remote?

Yes:
- `NAT_DIR_EGRESS` - a request to a remote backend.
- `NAT_DIR_INGRESS` - a reply from a remote backend + other packets.

**Comment by pchaigno** [comment]
> Ah, right. So if the backend is remote, I can just skip the ingress host policy enforcement. Is it correct to assume that if `CILIUM_CALL_IPV{4,6}_NODEPORT_NAT` is called with `NAT_DIR_EGRESS`, then the backend is remote? If that's always true, I could maybe set a mark to skip the host firewall (I'm guessing `ctx->tc_index` isn't preserved on redirects?).

@brb So my statement :point_up: was incorrect. We are redirected to the egress of whatever device is on the path to the remote backend. We will therefore hit `to-netdev` which enforces egress host policies. However, in this PR, we also have a commit (the one before last) that ensures the source ID for packets reaching `to-netdev` is resolved to `HOST_ID` only if the packet has the mark indicating it comes from the hostns. Therefore, packets coming from `nodeport_lb{4,6}` will be redirected to `to-netdev`, but they will bypass the host firewall because they don't have the `HOST_ID` mark. That is the expected behavior since those packets are not actually egressing the hostns.

I validated this with a modified NodePort e2e test (which I'm planning to add in a follow up PR).

**Comment by brb** [comment]
> We are redirected to the egress of whatever device is on the path to the remote backend. We will therefore hit to-netdev which enforces egress host policies.

@pchaigno Hmm, but in that case the packet has been already SNAT'd. Isn't it a problem?

**Comment by pchaigno** [comment]
> > We are redirected to the egress of whatever device is on the path to the remote backend. We will therefore hit to-netdev which enforces egress host policies.
> 
> @pchaigno Hmm, but in that case the packet has been already SNAT'd. Isn't it a problem?

@brb I don't think so because we won't rely on the source IP to determine if we should enforce egress host policies. We will rely on the packet mark. That packet won't have the mark unless it's coming from the hostns.

**Comment by brb** [comment]
> I don't think so because we won't rely on the source IP to determine if we should enforce egress host policies. 

@pchaigno OK, makes sense.

I'm not very familiar with the host-fw semantics, but what's about a use case in which a NodePort svc backends run in a host netns (currently broken, https://github.com/cilium/cilium/pull/10744 will fix that)? In the case of a remote backend, you probably still want to enforce the policy before the SNAT, don't you?

**Comment by pchaigno** [comment]
test-me-please

File: bpf/bpf_host.c
Comment (joestringer) [inline bpf/bpf_host.c]: I always figured that pinging a service IP should just work, although I'm not specifically aware of any requirement for it to do so. The commit message asserts that it doesn't work, do you know why?

File: bpf/bpf_host.c
Comment (joestringer) [inline bpf/bpf_host.c:180]: My other thought was about whether such ICMPv6 messages are subject to firewalling, but I note that in `icmp6_host_handle()` there are static rules for ICMPv6 firewall handling from PR #12049. :+1:

File: test/k8sT/Conformance.go
Comment (joestringer) [inline test/k8sT/Conformance.go]: Is this fixable? I'm not sure I fully understand the issue here. Wondering if it could affect someone and if so, whether it's worth at least tracking in a separate GH issue (which we could point to in these comments).

File: pkg/node/manager/manager.go
Comment (joestringer) [inline pkg/node/manager/manager.go:347]: Is there also a masquerading aspect to this?

My high level expectation (and I feel like we already discussed this so apologies in advance but hopefully we get a description for the future) is this: When traffic forwarded directly (not via tunnel) arrives at the destination node, we would use the ipcache to look up the source IP to map to an identity. If this doesn't work, then why?

(The commit message asserts that we must send such traffic via tunnel, but doesn't seem to include the underlying motivation or which features force this requirement).

File: bpf/bpf_lxc.c
Comment (joestringer) [inline bpf/bpf_lxc.c]: nit (commit message): `destined to the its host` -> `destined to the local host`?

File: bpf/bpf_lxc.c
Comment (joestringer) [inline bpf/bpf_lxc.c]: Do we need to capture the full packet here, or is `cilium_dbg()` sufficient?

Looks like this code was already here, but I'm not sure I understand what the `cilium_dbg_capture()` provides that `send_trace_notify()` doesn't. Seems like we should be able to trim back these extra dbg_captures.

File: bpf/bpf_lxc.c
Comment (joestringer) [inline bpf/bpf_lxc.c:341]: nit for readability: We could probably avoid duplicating the trace + redirect by splitting the logic:

```
if (is_defined(HOST_REDIRECT_TO_INGRESS)) {
   ... perform IPv6 l3
   redirect = true
} else if (is_defined(ENABLE_HOST_FIREWALL) && *dstID == HOST_ID) {
   redirect = true
}
if (redirect) {
   ....send_trace_notify() / return redirect()
}
```

(Not a big deal but I had to stare at this for a while to convince myself that the various parameters to the trace/redirect are identical)

File: bpf/bpf_lxc.c
Comment (joestringer) [inline bpf/bpf_lxc.c:704]: Equivalent points to the last couple of messages here.

File: bpf/bpf_host.c
Comment (joestringer) [reply bpf/bpf_host.c:180]: Ah, a later commit actually extends this so :+1:

File: examples/policies/host/lock-down-dev-vms.yaml
Comment (joestringer) [inline examples/policies/host/lock-down-dev-vms.yaml:1]: Is this example automagically picked up by the CI somehow or were you testing this just locally?

Would be awesome to hook it in as part of regular tests to prevent breakage. (But I also recognize that could involve more changes which may/may not make sense for backport).

**Review by joestringer — APPROVED** [review]
Nice work, well segmented into logical commits which made review much easier :pray:  :tada: 

I think all of my questions / nits here can be followed up on subsequently. There's a couple of instances where there are motivations which I didn't find clear and places where we could insert "GH-XXXXX: ____" comments to link the caveats to issues which would help tracking if someone needs background and particularly if any of the issues fell off our radar. Other than that just the regular nits which we can quibble about but I don't see as crucial to getting these fixes in.

File: examples/policies/host/lock-down-dev-vms.yaml
Comment (joestringer) [reply examples/policies/host/lock-down-dev-vms.yaml:1]: I guess this question is largely already addressed by your PR description above so I'll await your followups :-)

File: pkg/node/manager/manager.go
Comment (pchaigno) [reply pkg/node/manager/manager.go:347]: Yep, we need to send through tunnel because otherwise there's some SNAT happening and we lose the source security ID. I'll update the commit message.

Is it possible/sensible for users to run in tunneling mode even though direct routing to remote nodes works? I kind of expected that traffic from pods to remote nodes would be SNATed all the time; otherwise, why use a tunnel?

File: test/k8sT/Conformance.go
Comment (pchaigno) [reply test/k8sT/Conformance.go]: I'm not sure what the underlying issue is yet. If I remember correctly, it was some kind of issue with masquerading.

I was planning to file issues once this PR is merged (see https://github.com/cilium/cilium/pull/12345#discussion_r450008336), but it makes sense to do it now to include the reference in the test :+1:

File: bpf/bpf_host.c
Comment (pchaigno) [reply bpf/bpf_host.c]: You had me doubt, so I raise the question on #sig-datapath: https://cilium.slack.com/archives/CDKG8NNHK/p1594837445242500?thread_ts=1594837352.242000&cid=CDKG8NNHK. TL;DR. ICMP isn't expected to work for services, and according to the `nodeport_lb6` code, it won't.

File: bpf/bpf_lxc.c
Comment (pchaigno) [reply bpf/bpf_lxc.c]: Yep, I think you're right. I'll remove the `cilium_dbg_capture`.

File: pkg/node/manager/manager.go
Comment (joestringer) [reply pkg/node/manager/manager.go:347]: :+1: good to declare assumptions.

I guess I'd come from a perspective of "if you run a tunnel you know it 'works' anywhere regardless of the underlying fabric" and that "SNAT should be done as little as possible" but yeah probably if the user knows they're running all nodes in the same L2 or if they know that their underlying L3 fabric provides routing for the addresses (or run something like kube-router / metallb / bird for advertising the pod prefix) then they should just run direct routing. And I think your expectation is probably fair, if we don't have these prior guarantees then SNAT is the best bet we have to ensure that traffic arrives at a destination.

Anyway I think just documenting this in the commits/code and filing an issue for any caveats/remaining bugs is fine to address this in the context of this PR.

**Comment by joestringer** [comment]
Everything except GKE is passing so assuming that the only subsequent changes are minor docs / comments fixups or commit messages, we should be good to merge. Will defer to @pchaigno to follow up with @cilium/janitors to get this in.

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-focus K8sDatapathConfig MonitorAggregation Checks that monitor aggregation restricts notifications

**Comment by pchaigno** [comment]
test-me-please

**Comment by joestringer** [comment]
BPF checks / GKE are the only failures, BPF checks is false positive and GKE is not required. Merging.

**Pull request merged**
